### PR TITLE
Implement ActiveSpanSource support.

### PR DESCRIPTION
**Pull request by carlosalberto**
This is based on the recent ActiveSpanSource work in the Java repo (115 PR). Notes on this take:

1. BaseSpan kept the former Span calls, minus the tracer property and context manager methods. Span implements `finish()`, the respective actual context manager methods, and `context` and `tracer` properties. ActiveSpan implements `deactivate()`, the respective actual context manager methods, and `wrapper` property, for the wrapped Span.

2. ActiveSpanContinuation remained as its own class, not within the ActiveSpan class - inner classes, public-wise, are not as used in Python.

3. Added `start_active_span()` method only, while keeping the `start_span()` around without deprecating it, so that it the latter will keep its previous, ‘manual’ behaviour. As the Python API does not use a Builder approach, and uses a few optional parameters on these methods, it *felt* like it would be overdoing by providing three methods for the start span logic (the two new ones, and the existing, deprecated one).  Feedback on this?

4. Ignore active span parameter for the start() methods - should it default to False? Or to None, as the other optional methods do, and let the implementers decide?

5. Included a thread-local implementation as a way of sample/testing the current approach.

Any feedback or suggestion is welcome. Open to questions or any discussion.

### CODE REVIEW

File: opentracing/span.py [hunk 1]
```diff
@@ -210,21 +244,166 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+
+        IMPORTANT: the caller MUST call ActiveSpanContinuation.activate() and
+        call ActiveSpan.deactivate() or the associated Span will never be
+        finished. That is, calling capture() increments a refcount that must be
+        decremented somewhere else.
+        """
+        return self._noop_continuation
+
+    @property
+    def wrapped(self):
```
Comment (bhs) [inline opentracing/span.py]: why is this needed? (it's not there in OT-Java)

File: opentracing/tracer.py [hunk 2]
```diff
@@ -46,25 +79,42 @@ def start_span(self,
                    child_of=None,
                    references=None,
                    tags=None,
-                   start_time=None):
-        """Starts and returns a new Span representing a unit of work.
+                   start_time=None,
+                   ignore_active_span=False):
+        """Starts and returns a new Span representing a unit of work,
```
Comment (bhs) [inline opentracing/tracer.py]: I'd like to make sure that `start_active_span` seems like the default / preferred option, but I don't want us to break code that relies on `start_span` now... maybe something like `start_manual_span` to follow Java's `startManual()`?

File: opentracing/span.py (hunk 1 above)
Comment (carlosalberto) [inline opentracing/span.py]: Correct, this is not in the Java implementation - but I took the route the python's `Span` code takes, by defining properties `tracer` and `context` (`tracer` doesn't exist in the Java's interfaces, for example).

In theory, an `ActiveSpan` should always be wrapping a `Span` object - but I can remove it if you feel this `wrapped` property doesn't belong to the base hierarchy :)

File: opentracing/tracer.py (hunk 2 above)
Comment (carlosalberto) [inline opentracing/tracer.py]: Oh, yes, we can do that. Was wondering myself about that - so, in this case, I will add a new `start_manual_span` method AND mark the `start_span` as deprecated. Will put that update into the next iteration.

**Comment by carlosalberto** [comment]
Answering's @bhs question about how this works with frameworks doing async things - I tried with `gevent`, and it works by leveraging its `gevent.local` data just fine (which is mostly used through patching, i.e. patch `threading.local` with `gevent.monkey.patch_threading()` and have everything works just fine). With other frameworks it works similarly (i.e. `werkzug` has its context-local storage layer too).

I had been checking `asyncio` recently, and that one is something I'm still not sure about - thing is: the people behind it *DONT* want to add some kind of context-local fields/data, so we would need to take another route there. But haven't been able to find anything that could work for us there (yet).

Observe this is is still a relatively new API (i.e. Python 3.5 and newer). I'm wondering how this could change, too. And I definitely would love to take some time this week to take a deeper look into that.

**Comment by bhs** [comment]
@carlosalberto thanks. I would like to hear direct feedback from @clutchski and his team, as they have been [thinking about this from the other direction](https://gitter.im/opentracing/public?at=59232fa9631b8e4e61008e88).

**Comment by bhs** [comment]
(See https://github.com/opentracing/opentracing-python/pull/52 which incorporates some of these ideas)

**Comment by carlosalberto** [comment]
Oh yes. Forgot to close this PR. Had thought that removing my own branch would be enough ;)
